File: JavaScript/2458.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

let N, M, arr;

const bfs = (start) => {
  const visited = new Array(N).fill(0);
  const queue = [];
  visited[start] = 1;
  queue.push(start);

  while (queue.length > 0) {
    const cur = queue.shift();
    for (const i of arr[cur]) {
      if (!visited[i]) {
        visited[i] = 1;
        queue.push(i);
      }
    }
  }

  return visited;
};

const input = [];
rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  [N, M] = input.shift().split(' ').map(Number);
  arr = new Array(N).fill().map(() => []);

  for (let i = 0; i < M; i++) {
    const [a, b] = input[i].split(' ').map((x) => parseInt(x) - 1);
    arr[a].push(b);
  }

  const ans = [];
  for (let i = 0; i < N; i++) {
    ans.push(bfs(i));
  }

  let cnt = 0;
  for (let i = 0; i < ans.length; i++) {
    let flag = true;
    for (let j = 0; j < ans[i].length; j++) {
      if (ans[i][j] === 0 && ans[j][i] === 0) {
        flag = false;
        break;
      }
    }
    if (flag) cnt++;
  }

  console.log(cnt);
  rl.close();
});

File: Python/2458.py
import sys
input = sys.stdin.readline
from collections import deque
def bfs(start):
    visited = [0 for _ in range(N)]
    queue = deque()
    visited[start] = 1
    queue.append(start)
    while queue:
        cur = queue.popleft()
        for i in range(len(arr[cur])):
            if not visited[arr[cur][i]]:
                visited[arr[cur][i]] = 1
                queue.append(arr[cur][i])
    return visited
N, M = map(int, input().split())
arr = [[] for _ in range(N)]
for _ in range(M):
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    arr[a].append(b)
ans = []
for i in range(len(arr)):
    res = bfs(i)
    ans.append(res)
cnt = 0
for i in range(len(ans)):
    flag = True
    for j in range(len(ans[i])):
        if ans[i][j] == 0 and ans[j][i] == 0:
                flag = False
    if flag: cnt += 1
print(cnt)
